In [1]:
import numpy as np 
import pandas as pd 
import matplotlib.pyplot as plt
import matplotlib
import warnings
warnings.filterwarnings('ignore')

cust_df = pd.read_csv("./santander.csv", encoding='latin-1')
print('dataset shape:', cust_df.shape)
cust_df.head(3)

dataset shape: (76020, 371)


ID  var3  var15  imp_ent_var16_ult1  imp_op_var39_comer_ult1  \
0   1     2     23                 0.0                      0.0   
1   3     2     34                 0.0                      0.0   
2   4     2     23                 0.0                      0.0   

   imp_op_var39_comer_ult3  imp_op_var40_comer_ult1  imp_op_var40_comer_ult3  \
0                      0.0                      0.0                      0.0   
1                      0.0                      0.0                      0.0   
2                      0.0                      0.0                      0.0   

   imp_op_var40_efect_ult1  imp_op_var40_efect_ult3  ...  \
0                      0.0                      0.0  ...   
1                      0.0                      0.0  ...   
2                      0.0                      0.0  ...   

   saldo_medio_var33_hace2  saldo_medio_var33_hace3  saldo_medio_var33_ult1  \
0                      0.0                      0.0                     0.0   
1                      0.0                      0.0                     0.0   
2                      0.0                      0.0                     0.0   

   saldo_medio_var33_ult3  saldo_medio_var44_hace2  saldo_medio_var44_hace3  \
0                     0.0                      0.0                      0.0   
1                     0.0                      0.0                      0.0   
2                     0.0                      0.0                      0.0   

   saldo_medio_var44_ult1  saldo_medio_var44_ult3     var38  TARGET  
0                     0.0                     0.0  39205.17       0  
1                     0.0                     0.0  49278.03       0  
2                     0.0                     0.0  67333.77       0  

[3 rows x 371 columns]

산탄데르 은행 고객 만족도조사 결과
피처 - 익명화, 컬럼명 익명화, 수치, 0인 데이터가 대다수 - 희소성
타겟 - TARGET

# 데이터 탐색

In [2]:
cust_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 76020 entries, 0 to 76019
Columns: 371 entries, ID to TARGET
dtypes: float64(111), int64(260)
memory usage: 215.2 MB


100명중 4명이 불만족하는 데이터

In [2]:
cust_df['TARGET'].value_counts() #0 만족, 1 불만족
unsatisfied_cnt = cust_df[cust_df['TARGET']==1].TARGET.count()
total_cnt = cust_df.TARGET.count()

print(f'불만족 비율은 : {unsatisfied_cnt/total_cnt:.2f}')

불만족 비율은 : 0.04


최소값, Q1, Q2, Q3 가 0인 피처가 많다.
최소값 -999999.000000 인 피처 var3 > Nan > 중위값/최빈값 2로 대체

In [4]:
cust_df.describe()

ID           var3         var15  imp_ent_var16_ult1  \
count   76020.000000   76020.000000  76020.000000        76020.000000   
mean    75964.050723   -1523.199277     33.212865           86.208265   
std     43781.947379   39033.462364     12.956486         1614.757313   
min         1.000000 -999999.000000      5.000000            0.000000   
25%     38104.750000       2.000000     23.000000            0.000000   
50%     76043.000000       2.000000     28.000000            0.000000   
75%    113748.750000       2.000000     40.000000            0.000000   
max    151838.000000     238.000000    105.000000       210000.000000   

       imp_op_var39_comer_ult1  imp_op_var39_comer_ult3  \
count             76020.000000             76020.000000   
mean                 72.363067               119.529632   
std                 339.315831               546.266294   
min                   0.000000                 0.000000   
25%                   0.000000                 0.000000   
50%                   0.000000                 0.000000   
75%                   0.000000                 0.000000   
max               12888.030000             21024.810000   

       imp_op_var40_comer_ult1  imp_op_var40_comer_ult3  \
count             76020.000000             76020.000000   
mean                  3.559130                 6.472698   
std                  93.155749               153.737066   
min                   0.000000                 0.000000   
25%                   0.000000                 0.000000   
50%                   0.000000                 0.000000   
75%                   0.000000                 0.000000   
max                8237.820000             11073.570000   

       imp_op_var40_efect_ult1  imp_op_var40_efect_ult3  ...  \
count             76020.000000             76020.000000  ...   
mean                  0.412946                 0.567352  ...   
std                  30.604864                36.513513  ...   
min                   0.000000                 0.000000  ...   
25%                   0.000000                 0.000000  ...   
50%                   0.000000                 0.000000  ...   
75%                   0.000000                 0.000000  ...   
max                6600.000000              6600.000000  ...   

       saldo_medio_var33_hace2  saldo_medio_var33_hace3  \
count             76020.000000             76020.000000   
mean                  7.935824                 1.365146   
std                 455.887218               113.959637   
min                   0.000000                 0.000000   
25%                   0.000000                 0.000000   
50%                   0.000000                 0.000000   
75%                   0.000000                 0.000000   
max               50003.880000             20385.720000   

       saldo_medio_var33_ult1  saldo_medio_var33_ult3  \
count            76020.000000            76020.000000   
mean                12.215580                8.784074   
std                783.207399              538.439211   
min                  0.000000                0.000000   
25%                  0.000000                0.000000   
50%                  0.000000                0.000000   
75%                  0.000000                0.000000   
max             138831.630000            91778.730000   

       saldo_medio_var44_hace2  saldo_medio_var44_hace3  \
count             76020.000000             76020.000000   
mean                 31.505324                 1.858575   
std                2013.125393               147.786584   
min                   0.000000                 0.000000   
25%                   0.000000                 0.000000   
50%                   0.000000                 0.000000   
75%                   0.000000                 0.000000   
max              438329.220000             24650.010000   

       saldo_medio_var44_ult1  saldo_medio_var44_ult3         var38  \
count            76020.000000            76020.000000  7.602000e+04   

In [5]:
cust_df.var3.value_counts()

 2         74165
 8           138
-999999      116
 9           110
 3           108
           ...  
 231           1
 188           1
 168           1
 135           1
 87            1
Name: var3, Length: 208, dtype: int64

## 피처 선택 - id 삭제, var3 은 최빈값 2로 대체

In [3]:
cust_df.drop('ID', axis=1, inplace=True)
cust_df['var3'].replace(-999999, 2, inplace=True)

cust_df.head()

var3  var15  imp_ent_var16_ult1  imp_op_var39_comer_ult1  \
0     2     23                 0.0                      0.0   
1     2     34                 0.0                      0.0   
2     2     23                 0.0                      0.0   
3     2     37                 0.0                    195.0   
4     2     39                 0.0                      0.0   

   imp_op_var39_comer_ult3  imp_op_var40_comer_ult1  imp_op_var40_comer_ult3  \
0                      0.0                      0.0                      0.0   
1                      0.0                      0.0                      0.0   
2                      0.0                      0.0                      0.0   
3                    195.0                      0.0                      0.0   
4                      0.0                      0.0                      0.0   

   imp_op_var40_efect_ult1  imp_op_var40_efect_ult3  imp_op_var40_ult1  ...  \
0                      0.0                      0.0                0.0  ...   
1                      0.0                      0.0                0.0  ...   
2                      0.0                      0.0                0.0  ...   
3                      0.0                      0.0                0.0  ...   
4                      0.0                      0.0                0.0  ...   

   saldo_medio_var33_hace2  saldo_medio_var33_hace3  saldo_medio_var33_ult1  \
0                      0.0                      0.0                     0.0   
1                      0.0                      0.0                     0.0   
2                      0.0                      0.0                     0.0   
3                      0.0                      0.0                     0.0   
4                      0.0                      0.0                     0.0   

   saldo_medio_var33_ult3  saldo_medio_var44_hace2  saldo_medio_var44_hace3  \
0                     0.0                      0.0                      0.0   
1                     0.0                      0.0                      0.0   
2                     0.0                      0.0                      0.0   
3                     0.0                      0.0                      0.0   
4                     0.0                      0.0                      0.0   

   saldo_medio_var44_ult1  saldo_medio_var44_ult3          var38  TARGET  
0                     0.0                     0.0   39205.170000       0  
1                     0.0                     0.0   49278.030000       0  
2                     0.0                     0.0   67333.770000       0  
3                     0.0                     0.0   64007.970000       0  
4                     0.0                     0.0  117310.979016       0  

[5 rows x 370 columns]

## X, y 데이터 분할 


In [4]:
X_features = cust_df.iloc[:, :-1]
y_labels = cust_df.iloc[:, -1]
X_features.shape, y_labels.shape

((76020, 369), (76020,))

## 훈련/테스트 분할

In [5]:
from sklearn.model_selection import train_test_split

In [6]:
X_train, X_test, y_train, y_test = train_test_split(X_features, y_labels, test_size=0.2, random_state=0)
#X_train.count()
print(f'학습 데이터 shape : {X_train.shape}, 테스트 데이터 shape : {X_test.shape}')

train_cnt = y_train.count()
test_cnt = y_test.count()
print(f'학습 데이터 값의 비율 : {y_train.value_counts()/train_cnt}')
print(f'테스트 데이터  값의 비율 : {y_test.value_counts()/test_cnt}')

학습 데이터 shape : (60816, 369), 테스트 데이터 shape : (15204, 369)
학습 데이터 값의 비율 : 0    0.960964
1    0.039036
Name: TARGET, dtype: float64
테스트 데이터  값의 비율 : 0    0.9583
1    0.0417
Name: TARGET, dtype: float64


# DT 예측기

# RF 예측기

# XGBoost 예측기 p271

## 훈련 데이터 > 훈련:검증 분할

In [7]:
X_tr, X_val, y_tr, y_val = train_test_split(X_train, y_train, test_size=0.3, random_state=0)

## XGBoost model

In [8]:
%%time

from xgboost import XGBClassifier
from sklearn.metrics import roc_auc_score

xgb_clf = XGBClassifier(n_estimator=500, learning_rate=0.05, random_state=156)
xgb_clf.fit(X_tr, y_tr, early_stopping_rounds=100, eval_metric='auc', eval_set=[(X_tr, y_tr), (X_val, y_val)])
xgb_roc_score = roc_auc_score(y_test, xgb_clf.predict_proba(X_test)[:, 1])
xgb_roc_score

[09:26:12] WARNING: D:\bld\xgboost-split_1637426510059\work\src\learner.cc:576: 
Parameters: { "n_estimator" } might not be used.

  This could be a false alarm, with some parameters getting used by language bindings but
  then being mistakenly passed down to XGBoost core, or some parameter actually being used
  but getting flagged wrongly here. Please open an issue if you find any such cases.


[0]	validation_0-auc:0.82179	validation_1-auc:0.80068
[1]	validation_0-auc:0.83092	validation_1-auc:0.80941
[2]	validation_0-auc:0.83207	validation_1-auc:0.80903
[3]	validation_0-auc:0.83288	validation_1-auc:0.80889
[4]	validation_0-auc:0.83414	validation_1-auc:0.80924
[5]	validation_0-auc:0.83524	validation_1-auc:0.80907
[6]	validation_0-auc:0.83568	validation_1-auc:0.81005
[7]	validation_0-auc:0.83741	validation_1-auc:0.81088
[8]	validation_0-auc:0.83896	validation_1-auc:0.81305
[9]	validation_0-auc:0.83949	validation_1-auc:0.81363
[10]	validation_0-auc:0.83908	validation_1-auc:0.81277
[11]	v

0.8397812474965844

## 하이퍼오피티 피라미터 평가

In [9]:
from hyperopt.pyll.base import scope
from hyperopt import hp

search_space = {'n_estimators': scope.int(hp.quniform('n_estimators', 50, 300, 10)), 
                'max_depth': scope.int(hp.quniform('max_depth', 3, 10, 1)),
                'learning_rate': hp.loguniform('learning_rate', np.log(0.01), np.log(0.3)),
                'subsample' : hp.uniform('subsample', 0.5, 1.0),
                'colsample_bytree' : hp.uniform('colsample_bytree', 0.5, 1.0)
                }

In [10]:
from hyperopt import fmin, tpe, Trials, STATUS_OK
from xgboost import XGBClassifier
from sklearn.model_selection import cross_val_score

# 목적 함수 
def objective_func_xgb(params) : 

    model = XGBClassifier(
        n_estimator = 100, #params['n_estimators'],
        max_depth = params['max_depth'],
        learning_rate = params['learning_rate'],
        #subsample = params['subsample'],
        colsample_bytree = params['colsample_bytree'],
        random_state= 42,
        eval_metric = 'logloss'
    )

    score_mean = cross_val_score(model, X_train, y_train, cv=5, scoring='accuracy').mean()

    return {'loss': -1*score_mean, 'status': STATUS_OK}

In [16]:
trials = Trials()

best_params = fmin(fn = objective_func_xgb,
                   space = search_space,
                   algo = tpe.suggest,
                   max_evals = 50,
                   trials = trials
                   )

[14:37:43] WARNING: D:\bld\xgboost-split_1637426510059\work\src\learner.cc:576: 
Parameters: { "n_estimator" } might not be used.

  This could be a false alarm, with some parameters getting used by language bindings but
  then being mistakenly passed down to XGBoost core, or some parameter actually being used
  but getting flagged wrongly here. Please open an issue if you find any such cases.


[14:37:52] WARNING: D:\bld\xgboost-split_1637426510059\work\src\learner.cc:576: 
Parameters: { "n_estimator" } might not be used.

  This could be a false alarm, with some parameters getting used by language bindings but
  then being mistakenly passed down to XGBoost core, or some parameter actually being used
  but getting flagged wrongly here. Please open an issue if you find any such cases.


[14:38:01] WARNING: D:\bld\xgboost-split_1637426510059\work\src\learner.cc:576: 
Parameters: { "n_estimator" } might not be used.

  This could be a false alarm, with some parameters getting used by lan

In [17]:
best_params

{'colsample_bytree': 0.5915406502081058,
 'learning_rate': 0.09515274008275879,
 'max_depth': 5.0,
 'n_estimators': 80.0,
 'subsample': 0.8945558681046086}

In [18]:
best_model = XGBClassifier(n_estimators= int(best_params['n_estimators']), 
                           max_depth = int(best_params['max_depth']),
                           learning_rate=best_params['learning_rate'], 
                           subsample = best_params['subsample'],
                           colsample_bytree = best_params['colsample_bytree'],
                           random_state = 42,
                           eval_metric = 'logloss'
                           )

In [19]:
eval_list = [(X_tr, y_tr), (X_val, y_val)]

best_model.fit(X_tr, y_tr,
               early_stopping_rounds = 100,
               eval_set = eval_list,
               eval_metric = 'auc'
               )

[0]	validation_0-auc:0.71611	validation_1-auc:0.69878
[1]	validation_0-auc:0.75258	validation_1-auc:0.72856
[2]	validation_0-auc:0.75608	validation_1-auc:0.72912
[3]	validation_0-auc:0.75702	validation_1-auc:0.73017
[4]	validation_0-auc:0.80341	validation_1-auc:0.77683
[5]	validation_0-auc:0.79958	validation_1-auc:0.77106
[6]	validation_0-auc:0.81773	validation_1-auc:0.79137
[7]	validation_0-auc:0.81590	validation_1-auc:0.79066
[8]	validation_0-auc:0.81703	validation_1-auc:0.79134
[9]	validation_0-auc:0.82463	validation_1-auc:0.79867
[10]	validation_0-auc:0.82222	validation_1-auc:0.79587
[11]	validation_0-auc:0.82918	validation_1-auc:0.80361
[12]	validation_0-auc:0.82914	validation_1-auc:0.80371
[13]	validation_0-auc:0.83406	validation_1-auc:0.80946
[14]	validation_0-auc:0.83411	validation_1-auc:0.80826
[15]	validation_0-auc:0.83204	validation_1-auc:0.80605
[16]	validation_0-auc:0.83627	validation_1-auc:0.81010
[17]	validation_0-auc:0.83639	validation_1-auc:0.80941
[18]	validation_0-au

XGBClassifier(base_score=0.5, booster='gbtree', colsample_bylevel=1,
              colsample_bynode=1, colsample_bytree=0.5915406502081058,
              enable_categorical=False, eval_metric='logloss', gamma=0,
              gpu_id=-1, importance_type=None, interaction_constraints='',
              learning_rate=0.09515274008275879, max_delta_step=0, max_depth=5,
              min_child_weight=1, missing=nan, monotone_constraints='()',
              n_estimators=80, n_jobs=16, num_parallel_tree=1, predictor='auto',
              random_state=42, reg_alpha=0, reg_lambda=1, scale_pos_weight=1,
              subsample=0.8945558681046086, tree_method='exact',
              validate_parameters=1, verbosity=None)

In [20]:
from sklearn.metrics import roc_auc_score

xgb_roc_source = roc_auc_score(y_test, best_model.predict_proba(X_test)[:,1])
xgb_roc_source

0.8427807451896533

## 최적화 파라미터 기준 모델 평가

n_estimator=500  
learning_rate = 0.15  
max_depth = 5  


# LGBM

In [31]:
from lightgbm import LGBMClassifier
from sklearn.metrics import roc_auc_score

lgbm_clf = LGBMClassifier(n_estimator=500, learning_rate=0.05, random_state=156)
lgbm_clf.fit(X_tr, y_tr, early_stopping_rounds=100, eval_metric='auc', eval_set=[(X_tr, y_tr), (X_val, y_val)])
lgbm_roc_score = roc_auc_score(y_test, lgbm_clf.predict_proba(X_test)[:, 1])
lgbm_roc_score

[LightGBM] [Warning] Unknown parameter: n_estimator
[1]	training's auc: 0.82625	training's binary_logloss: 0.159747	valid_1's auc: 0.809814	valid_1's binary_logloss: 0.161487
[2]	training's auc: 0.83267	training's binary_logloss: 0.155995	valid_1's auc: 0.811143	valid_1's binary_logloss: 0.158489
[3]	training's auc: 0.835609	training's binary_logloss: 0.152875	valid_1's auc: 0.812893	valid_1's binary_logloss: 0.155925
[4]	training's auc: 0.837651	training's binary_logloss: 0.150255	valid_1's auc: 0.814154	valid_1's binary_logloss: 0.153849
[5]	training's auc: 0.840785	training's binary_logloss: 0.148052	valid_1's auc: 0.814868	valid_1's binary_logloss: 0.152054
[6]	training's auc: 0.844421	training's binary_logloss: 0.146074	valid_1's auc: 0.816352	valid_1's binary_logloss: 0.1505
[7]	training's auc: 0.846655	training's binary_logloss: 0.144325	valid_1's auc: 0.817164	valid_1's binary_logloss: 0.149162
[8]	training's auc: 0.848911	training's binary_logloss: 0.142724	valid_1's auc: 0.82

0.8403013083796488

## LGBM 하이퍼오피티 

In [11]:
from hyperopt.pyll.base import scope
from hyperopt import hp

lgbm_search_space = {'num_leaves': hp.quniform('num_leaves', 32, 64, 1),
                     'max_depth': hp.quniform('max_depth', 100, 160, 1),
                     'min_child_samples': hp.quniform('min_child_samples', 60, 100, 1),
                     'subsample': hp.uniform('subsample', 0.7, 1),
                     'learning_rate': hp.uniform('learning_rate', 0.01, 0.2)
                     }

In [23]:
from hyperopt import fmin, tpe, Trials, STATUS_OK
from lightgbm import LGBMClassifier
from sklearn.model_selection import cross_val_score

# 목적 함수 
def objective_func_lgbm(params) : 

    lgbm_clf = LGBMClassifier(
        n_estimator = 100, #params['n_estimators'],
        num_leaves= int(params['num_leaves']),
        max_depth = int(params['max_depth']),
        learning_rate = params['learning_rate'],
        subsample = params['subsample'],
        min_child_samples= int(params['min_child_samples'])
    )

    score_mean_lgbm = cross_val_score(lgbm_clf, X_train, y_train, cv=5, scoring='accuracy').mean()

    return {'loss': -1*score_mean_lgbm, 'status': STATUS_OK}

In [25]:
trials_lgbm = Trials()

best_params_lgbm = fmin(fn = objective_func_lgbm,
                   space = lgbm_search_space,
                   algo = tpe.suggest,
                   max_evals = 50,
                   trials = trials_lgbm
                   )

[LightGBM] [Warning] Unknown parameter: n_estimator   
[LightGBM] [Warning] Unknown parameter: n_estimator   
[LightGBM] [Warning] Unknown parameter: n_estimator   
[LightGBM] [Warning] Unknown parameter: n_estimator   
[LightGBM] [Warning] Unknown parameter: n_estimator                              
[LightGBM] [Warning] Unknown parameter: n_estimator                              
[LightGBM] [Warning] Unknown parameter: n_estimator                              
[LightGBM] [Warning] Unknown parameter: n_estimator                              
[LightGBM] [Warning] Unknown parameter: n_estimator                              
[LightGBM] [Warning] Unknown parameter: n_estimator                              
[LightGBM] [Warning] Unknown parameter: n_estimator                              
[LightGBM] [Warning] Unknown parameter: n_estimator                              
[LightGBM] [Warning] Unknown parameter: n_estimator                              
[LightGBM] [Warning] Unknown parameter: n_

In [26]:
best_params_lgbm

{'learning_rate': 0.02424490492343663,
 'max_depth': 128.0,
 'min_child_samples': 75.0,
 'num_leaves': 47.0,
 'subsample': 0.7788000808720014}

In [27]:
best_model_lgbm = LGBMClassifier(
        n_estimator = 100, #params['n_estimators'],
        num_leaves= int(best_params_lgbm['num_leaves']),
        max_depth = int(best_params_lgbm['max_depth']),
        learning_rate = best_params_lgbm['learning_rate'],
        subsample = best_params_lgbm['subsample'],
        min_child_samples= int(best_params_lgbm['min_child_samples'])
    )

In [28]:
eval_list_lgbm = [(X_tr, y_tr), (X_val, y_val)]

best_model_lgbm.fit(X_tr, y_tr,
               early_stopping_rounds = 100,
               eval_set = eval_list_lgbm,
               eval_metric = 'auc'
               )

[LightGBM] [Warning] Unknown parameter: n_estimator
[1]	training's auc: 0.832991	training's binary_logloss: 0.162128	valid_1's auc: 0.807032	valid_1's binary_logloss: 0.163489
[2]	training's auc: 0.835391	training's binary_logloss: 0.159987	valid_1's auc: 0.806258	valid_1's binary_logloss: 0.161731
[3]	training's auc: 0.84131	training's binary_logloss: 0.1581	valid_1's auc: 0.811051	valid_1's binary_logloss: 0.16015
[4]	training's auc: 0.84479	training's binary_logloss: 0.15635	valid_1's auc: 0.813204	valid_1's binary_logloss: 0.15872
[5]	training's auc: 0.84601	training's binary_logloss: 0.15478	valid_1's auc: 0.815348	valid_1's binary_logloss: 0.157418
[6]	training's auc: 0.846536	training's binary_logloss: 0.153308	valid_1's auc: 0.815948	valid_1's binary_logloss: 0.156281
[7]	training's auc: 0.846687	training's binary_logloss: 0.151937	valid_1's auc: 0.816294	valid_1's binary_logloss: 0.155212
[8]	training's auc: 0.847925	training's binary_logloss: 0.150654	valid_1's auc: 0.816374	

LGBMClassifier(learning_rate=0.02424490492343663, max_depth=128,
               min_child_samples=75, n_estimator=100, num_leaves=47,
               subsample=0.7788000808720014)

In [29]:
from sklearn.metrics import roc_auc_score

lgbm_roc_source = roc_auc_score(y_test, best_model_lgbm.predict_proba(X_test)[:,1])
lgbm_roc_source

0.8412117938203257

# XGRM vs LGBM
1. auc 기준 평가 결과
2. 각각의 파라미터 값 

xgb: 0.8397812474965844
xgb best: 0.8427807451896533
XGBClassifier(base_score=0.5, booster='gbtree', colsample_bylevel=1,
              colsample_bynode=1, colsample_bytree=0.5915406502081058,
              enable_categorical=False, eval_metric='logloss', gamma=0,
              gpu_id=-1, importance_type=None, interaction_constraints='',
              learning_rate=0.09515274008275879, max_delta_step=0, max_depth=5,
              min_child_weight=1, missing=nan, monotone_constraints='()',
              n_estimators=80, n_jobs=16, num_parallel_tree=1, predictor='auto',
              random_state=42, reg_alpha=0, reg_lambda=1, scale_pos_weight=1,
              subsample=0.8945558681046086, tree_method='exact',
              validate_parameters=1, verbosity=None)

lgbm: 0.8403013083796488
lgbm best: 0.8412117938203257
LGBMClassifier(learning_rate=0.02424490492343663, max_depth=128,
               min_child_samples=75, n_estimator=100, num_leaves=47,
               subsample=0.7788000808720014)